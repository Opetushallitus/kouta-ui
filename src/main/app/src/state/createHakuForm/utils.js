import get from 'lodash/get';
import pick from 'lodash/pick';

import { isNumeric } from '../../utils';
import { JULKAISUTILA } from '../../constants';
import { ErrorBuilder } from '../../validation';
import { getHakulomakeFieldsData, getHakulomakeFieldsValues } from '../utils';

const getKielivalinta = values => get(values, 'kieliversiot.languages') || [];

export const getHakuByValues = values => {
  const alkamiskausiKoodiUri = get(values, 'aikataulut.kausi') || null;

  const alkamisvuosi = isNumeric(get(values, 'aikataulut.vuosi.value'))
    ? parseInt(values.aikataulut.vuosi.value)
    : null;

  const kielivalinta = getKielivalinta(values);

  const hakutapaKoodiUri = get(values, 'hakutapa') || null;

  const hakuajat = (get(values, 'aikataulut.hakuaika') || []).map(
    ({ alkaa, paattyy }) => ({
      alkaa: alkaa || null,
      paattyy: paattyy || null,
    }),
  );

  const {
    hakulomaketyyppi,
    hakulomakeId,
    hakulomakeLinkki,
    hakulomakeKuvaus,
  } = getHakulomakeFieldsData({ values, kielivalinta });

  const hakukohteenLiittamisenTakaraja =
    get(values, 'aikataulut.lisaamisenTakaraja') || null;

  const ajastettuJulkaisu = get(values, 'aikataulut.ajastettuJulkaisu') || null;

  const nimi = pick(get(values, 'nimi.nimi') || null, kielivalinta);

  const kohdejoukkoKoodiUri = get(values, 'kohdejoukko.kohde') || null;

  const kohdejoukonTarkenneKoodiUri = null;

  const metadata = {
    tulevaisuudenAikataulu: (get(values, 'aikataulut.aikataulu') || []).map(
      ({ alkaa, paattyy }) => ({
        alkaa: alkaa || null,
        paattyy: paattyy || null,
      }),
    ),
    yhteyshenkilot: (get(values, 'yhteyshenkilot') || []).map(
      ({ nimi, titteli, puhelinnumero, sahkoposti, verkkosivu }) => ({
        nimi: pick(nimi || {}, kielivalinta),
        titteli: pick(titteli || {}, kielivalinta),
        puhelinnumero: pick(puhelinnumero || {}, kielivalinta),
        wwwSivu: pick(verkkosivu || {}, kielivalinta),
        sahkoposti: pick(sahkoposti || {}, kielivalinta),
      }),
    ),
  };

  const valintakokeet = (get(values, 'valintakoe.tyypit') || []).map(
    ({ value }) => ({
      tyyppi: value,
      tilaisuudet: (
        get(values, ['valintakoe', 'tilaisuudet', value]) || []
      ).map(
        ({
          osoite,
          postinumero,
          postitoimipaikka,
          alkaa,
          paattyy,
          lisatietoja,
        }) => ({
          osoite: {
            osoite: pick(osoite || {}, kielivalinta),
            postinumero: postinumero || null,
            postitoimipaikka: pick(postitoimipaikka || {}, kielivalinta),
          },
          aika: {
            alkaa: alkaa || null,
            paattyy: paattyy || null,
          },
          lisatietoja: pick(lisatietoja || {}, kielivalinta),
        }),
      ),
    }),
  );

  const hakukohteenMuokkaamisenTakaraja =
    get(values, 'aikataulut.muokkauksenTakaraja') || null;

  return {
    alkamiskausiKoodiUri,
    kielivalinta,
    hakutapaKoodiUri,
    hakuajat,
    hakukohteenLiittamisenTakaraja,
    nimi,
    kohdejoukkoKoodiUri,
    kohdejoukonTarkenneKoodiUri,
    hakulomaketyyppi,
    metadata,
    hakukohteenMuokkaamisenTakaraja,
    ajastettuJulkaisu,
    alkamisvuosi,
    hakulomakeId,
    hakulomakeLinkki,
    hakulomakeKuvaus,
    valintakokeet,
  };
};

export const getValuesByHaku = haku => {
  const {
    alkamiskausiKoodiUri = '',
    alkamisvuosi = '',
    hakuajat = [],
    hakutapaKoodiUri = '',
    kohdejoukkoKoodiUri = '',
    hakulomaketyyppi = '',
    hakulomakeId = '',
    hakulomakeKuvaus = {},
    hakulomakeLinkki = {},
    hakukohteenLiittamisenTakaraja,
    hakukohteenMuokkaamisenTakaraja,
    ajastettuJulkaisu,
    kielivalinta = [],
    valintakokeet = [],
    nimi = {},
    metadata = {},
  } = haku;

  const { tulevaisuudenAikataulu = [], yhteyshenkilot = [] } = metadata;

  return {
    nimi: {
      nimi: nimi,
    },
    kieliversiot: {
      languages: kielivalinta,
    },
    aikataulut: {
      kausi: alkamiskausiKoodiUri,
      vuosi: { value: alkamisvuosi ? alkamisvuosi.toString() : '' },
      hakuaika: (hakuajat || []).map(({ alkaa, paattyy }) => {
        return {
          alkaa,
          paattyy,
        };
      }),
      aikataulu: (tulevaisuudenAikataulu || []).map(({ alkaa, paattyy }) => {
        return {
          alkaa,
          paattyy,
        };
      }),
      lisaamisenTakaraja: hakukohteenLiittamisenTakaraja,
      muokkauksenTakaraja: hakukohteenMuokkaamisenTakaraja,
      ajastettuJulkaisu,
    },
    hakutapa: hakutapaKoodiUri,
    kohdejoukko: {
      kohde: kohdejoukkoKoodiUri,
    },
    hakulomake: getHakulomakeFieldsValues({
      hakulomaketyyppi,
      hakulomakeId,
      hakulomakeKuvaus,
      hakulomakeLinkki,
    }),
    yhteyshenkilot: yhteyshenkilot.map(
      ({ nimi, titteli, puhelinnumero, sahkoposti, wwwSivu }) => ({
        nimi: nimi || {},
        titteli: titteli || {},
        puhelinnumero: puhelinnumero || {},
        sahkoposti: sahkoposti || {},
        verkkosivu: wwwSivu || {},
      }),
    ),
    valintakoe: {
      tyypit: (valintakokeet || []).map(({ tyyppi }) => ({ value: tyyppi })),
      tilaisuudet: (valintakokeet || []).reduce(
        (acc, { tyyppi, tilaisuudet }) => ({
          ...acc,
          [tyyppi]: (tilaisuudet || []).map(
            ({ osoite, aika, lisatietoja }) => ({
              osoite: get(osoite, 'osoite') || {},
              postinumero: get(osoite, 'postinumero') || '',
              postitoimipaikka: get(osoite, 'postitoimipaikka') || {},
              alkaa: get(aika, 'alkaa') || '',
              paattyy: get(aika, 'paattyy') || '',
              lisatietoja: lisatietoja || {},
            }),
          ),
        }),
        {},
      ),
    },
  };
};

const validateEssentials = ({ values, errorBuilder }) => {
  const kielivalinta = getKielivalinta(values);

  return errorBuilder
    .validateArrayMinLength('kieliversiot.languages', 1)
    .validateTranslations('nimi.nimi', kielivalinta);
};

const validateCommon = ({ values, errorBuilder }) => {
  const hakutapa = get(values, 'hakutapa');
  const isYhteishaku = new RegExp('^hakutapa_01').test(hakutapa);
  const isErillishaku = new RegExp('^hakutapa_02').test(hakutapa);

  let enhancedErrorBuilder = errorBuilder
    .validateExistence('kohdejoukko.kohde')
    .validateExistence('hakutapa')
    .validateArrayMinLength('aikataulut.hakuaika', 1, { isFieldArray: true })
    .validateArray('aikataulut.hakuaika', eb => {
      if (isErillishaku || isYhteishaku) {
        return eb.validateExistence('alkaa').validateExistence('paattyy');
      }

      return eb.validateExistence('alkaa');
    });

  return enhancedErrorBuilder;
};

export const validate = ({ tila, values }) => {
  let errorBuilder = new ErrorBuilder({ values });

  errorBuilder = validateEssentials({ values, errorBuilder });

  if (tila === JULKAISUTILA.TALLENNETTU) {
    return errorBuilder.getErrors();
  }

  errorBuilder = validateCommon({ values, errorBuilder });

  return errorBuilder.getErrors();
};
