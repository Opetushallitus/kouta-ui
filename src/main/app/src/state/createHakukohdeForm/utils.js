import get from 'lodash/get';
import pick from 'lodash/pick';

import {
  isArray,
  getKoutaDateString,
  isNumeric,
  formatKoutaDateString,
} from '../../utils';

const DATE_FORMAT = 'DD.MM.YYYY HH:mm';

const getAsNumberOrNull = value => {
  return isNumeric(value) ? parseInt(value) : null;
};

const getDateTimeValues = dateString => {
  if (!dateString) {
    return { date: '', time: '' };
  }

  const formattedDate = formatKoutaDateString(dateString, DATE_FORMAT);

  const [d = '', t = ''] = formattedDate.split(' ');

  return { date: d, time: t };
};

const getKoutaDateStringByDateTime = ({ date = '', time = '' }) => {
  const [day, month, year] = date.split('.');

  if (!isNumeric(day) || !isNumeric(month) || !isNumeric(year)) {
    return null;
  }

  const [hour, minute] = time.split(':');

  return getKoutaDateString({
    day,
    month,
    year,
    hour: hour || 0,
    minute: minute || 0,
  });
};

export const getHakukohdeByValues = values => {
  const alkamiskausiKoodiUri = get(values, 'alkamiskausi.kausi') || null;
  const alkamisvuosi = getAsNumberOrNull(
    get(values, 'alkamiskausi.vuosi.value'),
  );
  const kielivalinta = get(values, 'kieliversiot.languages') || [];
  const aloituspaikat = getAsNumberOrNull(
    get(values, 'aloituspaikat.aloituspaikkamaara'),
  );
  const kaytetaanHaunAikataulua = !get(values, 'hakuajat.eriHakuaika');

  const hakuajat = kaytetaanHaunAikataulua
    ? []
    : (get(values, 'hakuajat.hakuajat') || []).map(
        ({ fromDate, fromTime, toDate, toTime }) => ({
          alkaa: getKoutaDateStringByDateTime({
            date: fromDate,
            time: fromTime,
          }),
          paattyy: getKoutaDateStringByDateTime({ date: toDate, time: toTime }),
        }),
      );

  const liitteidenToimitusosoite = {
    osoite: {
      osoite: pick(
        get(values, 'liitteet.toimitusosoite') || null,
        kielivalinta,
      ),
      postinumero: get(values, 'liitteet.toimituspostinumero') || null,
      postitoimipaikka: pick(
        get(values, 'liitteet.toimituspostitoimipaikka') || null,
        kielivalinta,
      ),
    },
    sahkoposti: get(values, 'liitteet.toimitussahkoposti') || null,
  };

  const liitteidenToimitusaika =
    get(values, 'liitteet.deliverDate') && get(values, 'liitteet.deliverTime')
      ? getKoutaDateStringByDateTime({
          date: values.liitteet.deliverDate,
          time: values.liitteet.deliverTime,
        })
      : null;

  const liitteetOnkoSamaToimitusosoite = !!get(
    values,
    'liitteet.yhteinenToimituspaikka',
  );

  const liitteetOnkoSamaToimitusaika = !!get(
    values,
    'liitteet.yhteinenToimitusaika',
  );

  const liitteet = (get(values, 'liitteet.liitteet') || []).map(
    ({
      tyyppi,
      nimi,
      kuvaus,
      deliverDate,
      deliverTime,
      toimitusosoite,
      toimituspostinumero,
      toimituspostitoimipaikka,
      toimitussahkoposti,
    }) => ({
      tyyppi: get(tyyppi, 'value') || null,
      nimi: pick(nimi || null, kielivalinta),
      toimitusaika: liitteetOnkoSamaToimitusaika
        ? getKoutaDateStringByDateTime({ date: deliverDate, time: deliverTime })
        : null,
      toimitusosoite: {
        osoite: {
          osoite: pick(toimitusosoite || null, kielivalinta),
          postinumero: toimituspostinumero || null,
          postitoimipaikka: pick(
            toimituspostitoimipaikka || null,
            kielivalinta,
          ),
        },
        sahkoposti: toimitussahkoposti || null,
      },
      kuvaus: pick(kuvaus || null, kielivalinta),
    }),
  );

  const nimi = pick(get(values, 'perustiedot.nimi') || null, kielivalinta);

  const toinenAsteOnkoKaksoistutkinto = !!get(
    values,
    'perustiedot.voiSuorittaaKaksoistutkinnon',
  );
  const valintakoeTyypit = get(values, 'valintakoe.types') || [];
  const valintakokeetByTyyppi = get(values, 'valintakoe.kokeet') || {};

  const valintakokeet = valintakoeTyypit
    .map(tyyppi => ({ tyyppi, ...(valintakokeetByTyyppi[tyyppi] || {}) }))
    .map(({ tyyppi, kokeet }) => ({
      tyyppi,
      tilaisuudet: isArray(kokeet)
        ? kokeet.map(
            ({
              osoite,
              postinumero,
              postitoimipaikka,
              fromDate,
              fromTime,
              toDate,
              toTime,
              lisatietoja,
            }) => ({
              osoite: {
                osoite: pick(osoite || null, kielivalinta),
                postinumero: postinumero || null,
                postitoimipaikka: pick(postitoimipaikka || null, kielivalinta),
              },
              aika: {
                alkaa: getKoutaDateStringByDateTime({
                  date: fromDate,
                  time: fromTime,
                }),
                paattyy: getKoutaDateStringByDateTime({
                  date: toDate,
                  time: toTime,
                }),
              },
              lisatietoja: pick(lisatietoja || null, kielivalinta),
            }),
          )
        : [],
    }));

  const pohjakoulutusvaatimusKoodiUrit = (
    get(values, 'pohjakoulutus.koulutusvaatimukset') || []
  ).map(({ value }) => value);

  const valintaperuste =
    get(values, 'valintaperusteenKuvaus.valintaperuste.value') || null;

  const ensikertalaisenAloituspaikat = getAsNumberOrNull(
    get(values, 'aloituspaikat.ensikertalaismaara'),
  );

  return {
    alkamiskausiKoodiUri,
    kaytetaanHaunAikataulua,
    kielivalinta,
    aloituspaikat,
    hakuajat,
    liitteetOnkoSamaToimitusaika,
    liitteetOnkoSamaToimitusosoite,
    liitteet,
    alkamisvuosi,
    liitteidenToimitusosoite,
    liitteidenToimitusaika,
    nimi,
    toinenAsteOnkoKaksoistutkinto,
    valintakokeet,
    pohjakoulutusvaatimusKoodiUrit,
    valintaperuste,
    ensikertalaisenAloituspaikat,
  };
};

export const getValuesByHakukohde = hakukohde => {
  const {
    alkamiskausiKoodiUri = '',
    kaytetaanHaunAikataulua,
    kielivalinta = [],
    aloituspaikat = '',
    hakuajat = [],
    liitteetOnkoSamaToimitusaika,
    liitteetOnkoSamaToimitusosoite,
    liitteet = [],
    alkamisvuosi,
    liitteidenToimitusosoite = {},
    liitteidenToimitusaika,
    nimi = {},
    toinenAsteOnkoKaksoistutkinto,
    valintakokeet = [],
    pohjakoulutusvaatimusKoodiUrit = [],
    valintaperuste = '',
    ensikertalaisenAloituspaikat = '',
  } = hakukohde;

  const valintakoeTyypit = (valintakokeet || []).map(({ tyyppi }) => tyyppi);

  const valintakokeetByTyyppi = (valintakokeet || []).reduce(
    (acc, { tyyppi, tilaisuudet = [] }) => {
      if (tyyppi) {
        acc[tyyppi] = {
          kokeet: (tilaisuudet || []).map(
            ({
              aika: { alkaa, paattyy } = {},
              osoite: {
                osoite = {},
                postinumero = '',
                postitoimipaikka = {},
              } = {},
              lisatietoja = {},
            }) => {
              const { date: fromDate, time: fromTime } = getDateTimeValues(
                alkaa,
              );

              const { date: toDate, time: toTime } = getDateTimeValues(paattyy);

              return {
                osoite,
                postinumero,
                postitoimipaikka,
                lisatietoja,
                fromDate,
                fromTime,
                toDate,
                toTime,
              };
            },
          ),
        };
      }

      return acc;
    },
    {},
  );

  const {
    date: liitteetDeliverDate,
    time: liitteetDeliverTime,
  } = getDateTimeValues(liitteidenToimitusaika);

  return {
    alkamiskausi: {
      kausi: alkamiskausiKoodiUri,
      vuosi: {
        value: isNumeric(alkamisvuosi) ? alkamisvuosi.toString() : null,
      },
    },
    kieliversiot: {
      languages: kielivalinta,
    },
    aloituspaikat: {
      aloituspaikkamaara: isNumeric(aloituspaikat)
        ? aloituspaikat.toString()
        : '',
      ensikertalaismaara: isNumeric(ensikertalaisenAloituspaikat)
        ? ensikertalaisenAloituspaikat.toString()
        : '',
    },
    hakuajat: {
      eriHakuaika: !kaytetaanHaunAikataulua,
      hakuajat: (hakuajat || []).map(({ alkaa, paattyy }) => {
        const { date: fromDate, time: fromTime } = getDateTimeValues(alkaa);
        const { date: toDate, time: toTime } = getDateTimeValues(paattyy);

        return {
          fromDate,
          fromTime,
          toDate,
          toTime,
        };
      }),
    },
    perustiedot: {
      nimi,
      voiSuorittaaKaksoistutkinnon: !!toinenAsteOnkoKaksoistutkinto,
    },
    pohjakoulutus: {
      koulutusvaatimus: (pohjakoulutusvaatimusKoodiUrit || []).map(value => ({
        value,
      })),
    },
    valintaperusteenKuvaus: {
      valintaperuste: {
        value: valintaperuste,
      },
    },
    valintakoe: {
      types: valintakoeTyypit,
      kokeet: valintakokeetByTyyppi,
    },
    liitteet: {
      toimitusosoite: get(liitteidenToimitusosoite, 'osoite.osoite') || {},
      toimituspostinumero:
        get(liitteidenToimitusosoite, 'osoite.postinumero') || '',
      toimituspostitoimipaikka:
        get(liitteidenToimitusosoite, 'osoite.postitoimipaikka') || {},
      toimitussahkoposti: get(liitteidenToimitusosoite, 'sahkoposti') || '',
      yhteinenToimituspaikka: !!liitteetOnkoSamaToimitusosoite,
      yhteinenToimitusaika: !!liitteetOnkoSamaToimitusaika,
      deliverDate: liitteetDeliverDate,
      deliverTime: liitteetDeliverTime,
      liitteet: (liitteet || []).map(
        ({
          tyyppi,
          nimi = {},
          toimitusaika,
          toimitusosoite: {
            sahkoposti,
            osoite: { osoite, postinumero, postitoimipaikka } = {},
          } = {},
          kuvaus = {},
        }) => {
          const { date: deliverDate, time: deliverTime } = getDateTimeValues(
            toimitusaika,
          );

          return {
            tyyppi,
            nimi,
            kuvaus,
            deliverDate,
            deliverTime,
            toimitussahkoposti: sahkoposti || '',
            toimitusosoite: osoite || {},
            toimituspostinumero: postinumero || '',
            toimituspostitoimipaikka: postitoimipaikka || {},
          };
        },
      ),
    },
  };
};
