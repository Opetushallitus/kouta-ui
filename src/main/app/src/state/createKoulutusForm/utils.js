import get from 'lodash/get';
import pick from 'lodash/pick';

import { JULKAISUTILA, KORKEAKOULUKOULUTUSTYYPIT } from '../../constants';
import { ErrorBuilder } from '../../validation';

const getKielivalinta = values => get(values, 'kieliversiot.languages') || [];

export const getKoulutusByValues = values => {
  const kielivalinta = getKielivalinta(values);
  const tarjoajat = get(values, 'tarjoajat') || [];
  const koulutusKoodiUri = get(values, 'information.koulutus.value') || null;
  const koulutustyyppi = get(values, 'type.type') || null;
  const osiot = get(values, 'lisatiedot.osiot') || [];
  const osioKuvaukset = get(values, 'lisatiedot.osioKuvaukset') || {};

  const osiotWithKuvaukset = osiot.map(({ value }) => ({
    otsikkoKoodiUri: value,
    teksti: pick(osioKuvaukset[value] || {}, kielivalinta),
  }));

  const kuvaus = pick(get(values, 'description.kuvaus') || {}, kielivalinta);
  const nimi = pick(get(values, 'information.nimi') || {}, kielivalinta);
  const opintojenLaajuusKoodiUri =
    get(values, 'information.opintojenLaajuus.value') || null;

  const tutkintonimikeKoodiUrit = (
    get(values, 'information.tutkintonimike') || []
  ).map(({ value }) => value);

  const koulutusalaKoodiUrit = (
    get(values, 'information.koulutusalat') || []
  ).map(({ value }) => value);

  const kuvauksenNimi = pick(
    get(values, 'description.nimi') || {},
    kielivalinta,
  );

  const julkinen = Boolean(get(values, 'nakyvyys.julkinen'));

  return {
    kielivalinta,
    tarjoajat,
    koulutusKoodiUri,
    koulutustyyppi,
    nimi,
    julkinen,
    metadata: {
      tyyppi: koulutustyyppi,
      lisatiedot: osiotWithKuvaukset,
      kuvaus,
      opintojenLaajuusKoodiUri,
      tutkintonimikeKoodiUrit,
      kuvauksenNimi,
      koulutusalaKoodiUrit,
    },
  };
};

export const getValuesByKoulutus = koulutus => {
  const {
    kielivalinta = [],
    koulutusKoodiUri = '',
    koulutustyyppi = '',
    tarjoajat = [],
    metadata = {},
    nimi = {},
    julkinen = false,
  } = koulutus;

  const {
    lisatiedot = [],
    kuvaus = {},
    opintojenLaajuusKoodiUri = '',
    tutkintonimikeKoodiUrit = [],
    kuvauksenNimi = {},
    koulutusalaKoodiUrit = [],
  } = metadata;

  const osiot = lisatiedot
    .filter(({ otsikkoKoodiUri }) => !!otsikkoKoodiUri)
    .map(({ otsikkoKoodiUri }) => ({ value: otsikkoKoodiUri }));

  const osioKuvaukset = lisatiedot.reduce((acc, curr) => {
    if (curr.otsikkoKoodiUri) {
      acc[curr.otsikkoKoodiUri] = curr.teksti || {};
    }

    return acc;
  }, {});

  return {
    kieliversiot: {
      languages: kielivalinta,
    },
    tarjoajat,
    information: {
      nimi,
      koulutus: {
        value: koulutusKoodiUri,
      },
      opintojenLaajuus: {
        value: opintojenLaajuusKoodiUri,
      },
      tutkintonimike: tutkintonimikeKoodiUrit.map(value => ({ value })),
      koulutusalat: koulutusalaKoodiUrit.map(value => ({ value })),
    },
    type: {
      type: koulutustyyppi,
    },
    lisatiedot: {
      osioKuvaukset,
      osiot,
    },
    description: {
      kuvaus,
      nimi: kuvauksenNimi,
    },
    nakyvyys: {
      julkinen,
    },
  };
};

const validateEssentials = ({ errorBuilder }) => {
  return errorBuilder
    .validateArrayMinLength('kieliversiot.languages', 1)
    .validateExistence('information.koulutus');
};

const validateCommon = ({ errorBuilder }) => {
  return errorBuilder.validateArrayMinLength('tarjoajat', 1);
};

const validateKorkeakoulu = ({ values, errorBuilder }) => {
  const kielivalinta = getKielivalinta(values);

  return errorBuilder.validateTranslations('description.kuvaus', kielivalinta);
};

export const validate = ({ values, koulutustyyppi, tila }) => {
  let errorBuilder = new ErrorBuilder({ values });

  errorBuilder = validateEssentials({ errorBuilder, values });

  if (tila === JULKAISUTILA.TALLENNETTU) {
    return errorBuilder.getErrors();
  }

  errorBuilder = validateCommon({ values, errorBuilder });

  if (KORKEAKOULUKOULUTUSTYYPIT.includes(koulutustyyppi)) {
    errorBuilder = validateKorkeakoulu({ values, errorBuilder });
  }

  return errorBuilder.getErrors();
};
