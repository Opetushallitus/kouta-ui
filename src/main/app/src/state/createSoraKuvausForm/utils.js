import get from 'lodash/get';
import pick from 'lodash/pick';
import mapValues from 'lodash/mapValues';

import { JULKAISUTILA } from '../../constants';
import createErrorBuilder from '../../utils/createErrorBuilder';
import parseEditorState from '../../utils/draft/parseEditorState';
import serializeEditorState from '../../utils/draft/serializeEditorState';

const getKielivalinta = values => get(values, 'kieliversiot') || [];

export const getSoraKuvausByValues = values => {
  const kielivalinta = getKielivalinta(values);
  const nimi = pick(get(values, 'tiedot.nimi') || {}, kielivalinta);
  const kuvaus = pick(get(values, 'tiedot.kuvaus') || {}, kielivalinta);
  const koulutustyyppi = get(values, 'koulutustyyppi') || null;
  const julkinen = Boolean(get(values, 'julkisuus'));

  return {
    nimi,
    julkinen,
    koulutustyyppi,
    kielivalinta,
    metadata: {
      kuvaus: mapValues(kuvaus, serializeEditorState),
    },
  };
};

export const getValuesBySoraKuvaus = soraKuvaus => {
  const {
    nimi,
    julkinen,
    koulutustyyppi,
    kielivalinta,
    metadata = {},
  } = soraKuvaus;

  const { kuvaus } = metadata;

  return {
    kieliversiot: kielivalinta || [],
    tiedot: {
      nimi: nimi || {},
      kuvaus: mapValues(kuvaus || {}, parseEditorState),
    },
    julkinen: Boolean(julkinen),
    koulutustyyppi: koulutustyyppi || null,
  };
};

const validateEssentials = ({ values, errorBuilder }) => {
  const kielivalinta = getKielivalinta(values);

  return errorBuilder
    .validateArrayMinLength('kieliversiot', 1)
    .validateTranslations('tiedot.nimi', kielivalinta);
};

const validateCommon = ({ values, errorBuilder }) => {
  const kielivalinta = getKielivalinta(values);

  return errorBuilder
    .validateTranslations('tiedot.kuvaus', kielivalinta)
    .validateExistence('koulutustyyppi');
};

export const validate = ({ tila, values }) => {
  let errorBuilder = createErrorBuilder(values);

  errorBuilder = validateEssentials({ values, errorBuilder });

  if (tila === JULKAISUTILA.TALLENNETTU) {
    return errorBuilder.getErrors();
  }

  errorBuilder = validateCommon({ values, errorBuilder });

  return errorBuilder.getErrors();
};
