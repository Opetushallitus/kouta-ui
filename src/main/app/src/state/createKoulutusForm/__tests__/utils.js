import { getKoulutusByValues, getValuesByKoulutus } from '../utils';

test('getKoulutusByValues returns correct koulutus given form values', () => {
  const koulutus = getKoulutusByValues({
    kieliversiot: ['fi', 'sv'],
    tarjoajat: ['123.456.789'],
    information: {
      nimi: {
        fi: 'Fi nimi',
        sv: 'Sv nimi',
      },
      koulutus: {
        value: 'koulutuskoodi_1#1',
      },
      opintojenLaajuus: {
        value: 'laajuus_1#1',
      },
      tutkintonimike: [{ value: 'nimike_1#1' }, { value: 'nimike_2#1' }],
      koulutusalat: [
        { value: 'koulutusala_1#1' },
        { value: 'koulutusala_2#1' },
      ],
    },
    koulutustyyppi: 'yo',
    lisatiedot: {
      osioKuvaukset: {
        'osio_1#1': {},
        'osio_2#1': {
          fi: 'Fi kuvaus',
          sv: 'Sv kuvaus',
        },
      },
      osiot: [{ value: 'osio_1#1' }, { value: 'osio_2#1' }],
    },
    description: {
      kuvaus: {
        fi: 'Fi kuvaus',
        sv: 'Sv kuvaus',
      },
      nimi: {
        fi: 'Fi nimi',
        sv: 'Sv nimi',
      },
    },
    julkinen: true,
  });

  expect(koulutus).toMatchSnapshot();
});

test('getValuesByKoulutus returns correct form values given koulutus', () => {
  const values = getValuesByKoulutus({
    julkinen: true,
    kielivalinta: ['fi', 'sv'],
    koulutusKoodiUri: 'koulutuskoodi_1#1',
    koulutustyyppi: 'yo',
    metadata: {
      kuvauksenNimi: {
        fi: 'Fi nimi',
        sv: 'Sv nimi',
      },
      kuvaus: {
        fi: 'Fi kuvaus',
        sv: 'Sv kuvaus',
      },
      lisatiedot: [
        {
          otsikkoKoodiUri: 'osio_1#1',
          teksti: {},
        },
        {
          otsikkoKoodiUri: 'osio_2#1',
          teksti: {
            fi: 'Fi kuvaus',
            sv: 'Sv kuvaus',
          },
        },
      ],
      opintojenLaajuusKoodiUri: 'laajuus_1#1',
      tutkintonimikeKoodiUrit: ['nimike_1#1', 'nimike_2#1'],
      tyyppi: 'yo',
      koulutusalaKoodiUrit: ['koulutusala_1#1', 'koulutusala_2#1'],
    },
    nimi: {
      fi: 'Fi nimi',
      sv: 'Sv nimi',
    },
    tarjoajat: ['123.456.789'],
  });

  expect(values).toMatchSnapshot();
});
