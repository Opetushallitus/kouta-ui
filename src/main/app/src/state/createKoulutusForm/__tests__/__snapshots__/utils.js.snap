// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getKoulutusByValues returns correct koulutus given form values 1`] = `
Object {
  "julkinen": true,
  "kielivalinta": Array [
    "fi",
    "sv",
  ],
  "koulutusKoodiUri": "koulutuskoodi_1#1",
  "koulutustyyppi": "yo",
  "metadata": Object {
    "koulutusalaKoodiUrit": Array [
      "koulutusala_1#1",
      "koulutusala_2#1",
    ],
    "kuvauksenNimi": Object {
      "fi": "Fi nimi",
      "sv": "Sv nimi",
    },
    "kuvaus": Object {
      "fi": "Fi kuvaus",
      "sv": "Sv kuvaus",
    },
    "lisatiedot": Array [
      Object {
        "otsikkoKoodiUri": "osio_1#1",
        "teksti": Object {},
      },
      Object {
        "otsikkoKoodiUri": "osio_2#1",
        "teksti": Object {
          "fi": "Fi kuvaus",
          "sv": "Sv kuvaus",
        },
      },
    ],
    "opintojenLaajuusKoodiUri": "laajuus_1#1",
    "tutkintonimikeKoodiUrit": Array [
      "nimike_1#1",
      "nimike_2#1",
    ],
    "tyyppi": "yo",
  },
  "nimi": Object {
    "fi": "Fi nimi",
    "sv": "Sv nimi",
  },
  "tarjoajat": Array [
    "123.456.789",
  ],
}
`;

exports[`getValuesByKoulutus returns correct form values given koulutus 1`] = `
Object {
  "description": Object {
    "kuvaus": Object {
      "fi": "Fi kuvaus",
      "sv": "Sv kuvaus",
    },
    "nimi": Object {
      "fi": "Fi nimi",
      "sv": "Sv nimi",
    },
  },
  "information": Object {
    "koulutus": Object {
      "value": "koulutuskoodi_1#1",
    },
    "koulutusalat": Array [
      Object {
        "value": "koulutusala_1#1",
      },
      Object {
        "value": "koulutusala_2#1",
      },
    ],
    "nimi": Object {
      "fi": "Fi nimi",
      "sv": "Sv nimi",
    },
    "opintojenLaajuus": Object {
      "value": "laajuus_1#1",
    },
    "tutkintonimike": Array [
      Object {
        "value": "nimike_1#1",
      },
      Object {
        "value": "nimike_2#1",
      },
    ],
  },
  "kieliversiot": Object {
    "languages": Array [
      "fi",
      "sv",
    ],
  },
  "lisatiedot": Object {
    "osioKuvaukset": Object {
      "osio_1#1": Object {},
      "osio_2#1": Object {
        "fi": "Fi kuvaus",
        "sv": "Sv kuvaus",
      },
    },
    "osiot": Array [
      Object {
        "value": "osio_1#1",
      },
      Object {
        "value": "osio_2#1",
      },
    ],
  },
  "nakyvyys": Object {
    "julkinen": true,
  },
  "organization": Object {
    "organizations": Array [
      "123.456.789",
    ],
  },
  "type": Object {
    "type": "yo",
  },
}
`;
